crimson/os/seastore/linked_tree_node: correct balance pivots calculation for both fixed-kv tree and omap tree

xxhdx1985126 opened the pull request:
Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>



<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins test classic perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-classic/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test crimson perf` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-perf-crimson/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-perf-pull-requests/config/definitions/ceph-perf-pull-requests.yml)
- `jenkins test signed` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pr-commits/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-commits/config/definitions/ceph-pr-commits.yml)
- `jenkins test make check` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests/config/definitions/ceph-pull-requests.yml)
- `jenkins test make check arm64` [Jenkins Job](https://jenkins.ceph.com/job/ceph-pull-requests-arm64/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pull-requests-arm64/config/definitions/ceph-pull-requests-arm64.yml)
- `jenkins test submodules` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-pr-submodules/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-submodules/config/definitions/ceph-pr-commits.yml)
- `jenkins test dashboard` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-pull-requests/config/definitions/ceph-dashboard-pull-requests.yml)
- `jenkins test dashboard cephadm` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-dashboard-cephadm-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-dashboard-cephadm-e2e/config/definitions/ceph-dashboard-cephadm-e2e.yml)
- `jenkins test api` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-api/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-pr-api/config/definitions/ceph-pr-api.yml)
- `jenkins test docs` [ReadTheDocs](https://readthedocs.org/projects/ceph/) | [Github Workflow Definition](https://github.com/ceph/ceph/blob/main/.readthedocs.yml)
- `jenkins test ceph-volume all` [Jenkins Jobs](https://jenkins.ceph.com/view/ceph-volume%20PR/) | [Jenkins Jobs Definition](https://github.com/ceph/ceph-build/blob/main/ceph-volume-cephadm-prs/config/definitions/ceph-volume-pr.yml)
- `jenkins test windows` [Jenkins Job](https://jenkins.ceph.com/job/ceph-windows-pull-requests/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-windows-pull-requests/config/definitions/ceph-windows-pull-requests.yml)
- `jenkins test rook e2e` [Jenkins Job](https://jenkins.ceph.com/view/all/job/ceph-orchestrator-rook-e2e/) | [Jenkins Job Definition](https://github.com/ceph/ceph-build/blob/main/ceph-rook-e2e/config/definitions/ceph-orchestrator-rook-e2e.yml)
</details>

cyx1231st commented on src/crimson/common/fixed_kv_node_layout.h:
```diff
@@ -558,6 +558,19 @@ class FixedKVNodeLayout {
+  static uint32_t get_balance_pivot_idx(
+    const FixedKVNodeLayout &left,
+    const FixedKVNodeLayout &right,
+    bool prefer_left)
```
I see in the fixed-btree, `prefer_left==false` means the removed item was in the right, else it was in the left.

But I can't recall why `pivot_idx` needs to be processed like this, do you remember why
```c++
if (total % 2 && prefer_left) {
  pivot_idx++;
}
```
?

xxhdx1985126 replied:
It seems that `prefer_left` indicates whether the donor of the balancing is from the left size. If it is, `prefer_left` is false; otherwise, it's true.

cyx1231st replied:
@athanatos  do you remember why this (`prefer_left` and `pivot_idx++`) is needed foundamentally?

athanatos replied:
The intent seems to that we keep the odd element on the node to which we are trying to balance elements (the one which just had a removal).  So, if the donor is on the left, we set prefer_left=false to put the extra on the right.  I can't actually remember whether this is strictly necessary, or simply that we expect more removals on the balance target.

athanatos replied:
I can't think of a reason why it would be required now -- as long as the donor has more than min putting the extra on either should still satisfy the invariant that both have at least min.  I might simply have misunderstood an edge case at the time.

xxhdx1985126 replied:
I've just went through `Introduction to algorithms, 3rd edition, Thomas H. Cormen`.

According to chapter 18.3, when deleting a key from a node, if the node has only `t - 1` keys and one of its siblings has `t` keys, the node should borrow one key from that sibling. So, following this, I think the `prefer_left` has its value in making sure to borrow one key from the right sibling if it has exacly `t` keys.

However, this does seem unnecessary as it just change the case from `t-1-key left, t-key right` to `t-key left, t-1-key right`. But I'm not sure if this is really unnecessary, it's textbook after all. So I'm throw this here.

xxhdx1985126 replied:
On the other hand, it seems that the current lba tree may inevitablly hit the case of removing a key from a minimal-occupancy node, since it handles the merge after removing the key. This may be violating the b+ tree invariant?

So I think maybe we should change it to handle the merge before removing the key. Does this make sense to you guys? @cyx1231st @athanatos

cyx1231st commented:
There is a potential issue related to linked omap tree https://tracker.ceph.com/issues/71307

cyx1231st commented:
For the https://tracker.ceph.com/issues/71307 omap issue that rebalance makes no difference, I think we can only:
1. Calculate correctly in advance and prevent useless rebalancing.
2. Or tolerate this in `linked_tree_node.h` if 1. is too difficult.

xxhdx1985126 commented:
@cyx1231st I've just modified the code as discussed, please take a look, thanks:-)

cyx1231st commented:
```
Start 273: unittest-fixed-kv-node-layout
220/317 Test #273: unittest-fixed-kv-node-layout .............***Exception: SegFault  0.01 sec
Running main() from gmock_main.cc
[==========] Running 6 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 6 tests from FixedKVNodeTest
[ RUN      ] FixedKVNodeTest.basic
[       OK ] FixedKVNodeTest.basic (0 ms)
[ RUN      ] FixedKVNodeTest.at_capacity
[       OK ] FixedKVNodeTest.at_capacity (0 ms)
[ RUN      ] FixedKVNodeTest.split
[       OK ] FixedKVNodeTest.split (0 ms)
[ RUN      ] FixedKVNodeTest.merge
[       OK ] FixedKVNodeTest.merge (0 ms)
[ RUN      ] FixedKVNodeTest.balanced
SEGFAULT
```

xxhdx1985126 commented:
> ```
> Start 273: unittest-fixed-kv-node-layout
> 220/317 Test #273: unittest-fixed-kv-node-layout .............***Exception: SegFault  0.01 sec
> Running main() from gmock_main.cc
> [==========] Running 6 tests from 1 test suite.
> [----------] Global test environment set-up.
> [----------] 6 tests from FixedKVNodeTest
> [ RUN      ] FixedKVNodeTest.basic
> [       OK ] FixedKVNodeTest.basic (0 ms)
> [ RUN      ] FixedKVNodeTest.at_capacity
> [       OK ] FixedKVNodeTest.at_capacity (0 ms)
> [ RUN      ] FixedKVNodeTest.split
> [       OK ] FixedKVNodeTest.split (0 ms)
> [ RUN      ] FixedKVNodeTest.merge
> [       OK ] FixedKVNodeTest.merge (0 ms)
> [ RUN      ] FixedKVNodeTest.balanced
> SEGFAULT
> ```

Looking into it:-)

cyx1231st commented on src/crimson/common/fixed_kv_node_layout.h:
```diff
@@ -558,36 +559,35 @@ class FixedKVNodeLayout {
-    }
-    return pivot_idx;
+    return (pivot_idx == l_size || pivot_idx == r_size)
+      ? std::nullopt
```
Is it really possible or reasonable hitting into this condition (nullopt) ?

See `fixed_tree::handle_merge()`, `merge_level()` is only called if the node is below min capacity (max-1)/2, this means, e.g.:
* cap = 6, min = 2, below-min = 1
* cap = 7, min = 3, below-min = 2

IIUC this means one of `left` and `right` must at below-min, then I don't understand why it is still possible to become `pivot_idx == l_size || pivot_idx == r_size`, because if the other node is at `min`, the node should be able to be merged instead, right ?

xxhdx1985126 replied:
There's a unit test case, `FixedKVNodeTest.balanced`, that can lead to this scenario. Should we remove that? Or keep the code? What do you think? @cyx1231st

cyx1231st replied:
Remove or simplify according to the current discussions, I suppose.

cyx1231st commented on src/crimson/common/fixed_kv_node_layout.h:
```diff
@@ -558,36 +559,35 @@ class FixedKVNodeLayout {
     set_meta(Meta::merge_from(left.get_meta(), right.get_meta()));
```
Sorry, it's in btree/fixed_kv_btree.h.

cyx1231st commented on src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:
```diff
@@ -786,14 +786,15 @@ class StringKVInnerNodeLayout {
     uint32_t right_size = omap_inner_key_t(right.get_node_key_ptr()[right.get_size()-1]).key_off;
     uint32_t total = left_size + right_size;
     uint32_t pivot_size = total / 2;
-    uint32_t pivot_idx = get_balance_pivot_idx(left, right);
```
IIUC we won't tolerate underflow nodes in the fix-kv-btrees.

So the commit message, e.g.:
```
crimson/os/seastore: adjust and fix rebalance logic in lba/backref/omap tree
* For fixed_kv_btree, drop `prefer_left`.
* For omap btree, tolerate underflow nodes if rebalance has the same results.

Fixes: https://tracker.ceph.com/issues/71307
```

cyx1231st replied:
Fixed.

cyx1231st commented on src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:
```diff
@@ -786,14 +786,15 @@ class StringKVInnerNodeLayout {
     StringKVInnerNodeLayout &replacement_left,
     StringKVInnerNodeLayout &replacement_right)
   {
+    ceph_assert(!left.below_min() || !right.below_min());
```
IIUC this can be parsed as: if both left and right are below min, they should be merged rather than rebalanced, right?

So,
```
ceph_assert(!(left.below_min() && right.below_min()));
```
can look more straightforward without documentation.

cyx1231st replied:
Was `ceph_assert(!(left.below_min() && !right.below_min()));`

cyx1231st reviewed (commented):
LGTM except for the comments.

xxhdx1985126 commented:
All done. Please take a look:-)

cyx1231st commented:
Changeset: addressed the above 2 comments

PTAL.

cyx1231st commented:
building at https://shaman.ceph.com/builds/ceph/wip-yingxin-testing-2025-05-19-1136

xxhdx1985126 commented:
> Changeset: addressed the above 2 comments in-place.
> 
> PTAL.

Ah, sorry, I missed the comments.

cyx1231st commented on src/crimson/common/fixed_kv_node_layout.h:
```diff
@@ -5,6 +5,7 @@
 
 #include <algorithm>
 #include <iostream>
+#include <optional>
```
Fixed.

cyx1231st commented:
building at https://shaman.ceph.com/builds/ceph/wip-yingxin-testing-2025-05-19-1159
